File: go/tracing/middleware.go
package tracing

import (
	"net/http"
	"regexp"
	"strings"
	"time"
)

// responseWriter wraps http.ResponseWriter to capture status code.
type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func newResponseWriter(w http.ResponseWriter) *responseWriter {
	return &responseWriter{w, http.StatusOK}
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

// Middleware returns a Chi-compatible middleware for distributed tracing.
// It extracts or generates trace_id from X-Trace-ID header and publishes
// start/end events for each request.
func Middleware(publisher *Publisher, service ServiceName) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Skip tracing for certain paths
			if shouldSkip(r.URL.Path) {
				next.ServeHTTP(w, r)
				return
			}

			// Extract or generate trace ID
			traceID := r.Header.Get(TraceIDHeader)
			if traceID == "" {
				traceID = GenerateTraceID()
			}

			// Set up trace context
			ctx := WithTraceContext(r.Context(), traceID)
			spanID := GetSpanID(ctx)

			// Build metadata
			metadata := map[string]any{
				"method":      r.Method,
				"path":        r.URL.Path,
				"query":       r.URL.RawQuery,
				"remote_addr": r.RemoteAddr,
				"user_agent":  r.UserAgent(),
			}

			// Build tags
			tags := map[string]string{
				"endpoint": normalizePath(r.URL.Path),
				"method":   r.Method,
			}

			// Publish start event
			publisher.PublishStart(ctx, Operations.HTTPRequestReceived, service, metadata, tags)
			startTime := time.Now()

			// Wrap response writer to capture status
			wrapped := newResponseWriter(w)

			// Add trace headers to response
			wrapped.Header().Set(TraceIDHeader, traceID)
			wrapped.Header().Set(SpanIDHeader, spanID)

			// Process request
			next.ServeHTTP(wrapped, r.WithContext(ctx))

			// Calculate duration
			durationMs := int(time.Since(startTime).Milliseconds())

			// Determine status
			status := StatusSuccess
			var errorMsg string
			if wrapped.statusCode >= 500 {
				status = StatusError
				errorMsg = http.StatusText(wrapped.statusCode)
			}

			// Update metadata and tags
			metadata["status_code"] = wrapped.statusCode
			metadata["duration_ms"] = durationMs
			tags["status_code"] = http.StatusText(wrapped.statusCode)

			// Publish end event
			publisher.PublishEnd(
				ctx,
				Operations.HTTPResponseSent,
				service,
				spanID,
				status,
				durationMs,
				metadata,
				tags,
				errorMsg,
			)
		})
	}
}

// skipPaths are paths that don't need tracing.
var skipPaths = []string{"/health", "/metrics", "/favicon.ico"}

func shouldSkip(path string) bool {
	for _, skip := range skipPaths {
		if strings.HasPrefix(path, skip) {
			return true
		}
	}
	return false
}

// uuidPattern matches UUID-like strings.
var uuidPattern = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

// numericPattern matches numeric IDs.
var numericPattern = regexp.MustCompile(`^\d+$`)

// normalizePath normalizes URL path for consistent tagging.
// /api/users/123 -> /api/users/{id}
// /api/tasks/abc-def-ghi -> /api/tasks/{id}
func normalizePath(path string) string {
	parts := strings.Split(path, "/")
	normalized := make([]string, 0, len(parts))

	for _, part := range parts {
		if part == "" {
			continue
		}
		// UUID-like or numeric ID
		if uuidPattern.MatchString(part) || numericPattern.MatchString(part) || len(part) > 20 {
			normalized = append(normalized, "{id}")
		} else {
			normalized = append(normalized, part)
		}
	}

	if len(normalized) == 0 {
		return "/"
	}
	return "/" + strings.Join(normalized, "/")
}

// InjectTraceHeaders adds trace context to outgoing HTTP request headers.
// Use when making requests to other services.
func InjectTraceHeaders(r *http.Request) *http.Request {
	ctx := r.Context()
	if traceID := GetTraceID(ctx); traceID != "" {
		r.Header.Set(TraceIDHeader, traceID)
	}
	if spanID := GetSpanID(ctx); spanID != "" {
		r.Header.Set(SpanIDHeader, spanID)
	}
	return r
}

// TracedClient wraps http.Client with automatic trace header injection.
type TracedClient struct {
	*http.Client
}

// NewTracedClient creates a new traced HTTP client.
func NewTracedClient(client *http.Client) *TracedClient {
	if client == nil {
		client = http.DefaultClient
	}
	return &TracedClient{client}
}

// Do performs the request with trace headers injected.
func (c *TracedClient) Do(r *http.Request) (*http.Response, error) {
	InjectTraceHeaders(r)
	return c.Client.Do(r)
}

File: python/jarvis_tracing/middleware.py
"""
Tracing Middleware
==================

FastAPI/Starlette middleware for automatic request tracing.
Extracts or generates trace_id from X-Trace-ID header.
"""

import time
from typing import Callable, Optional

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response
from starlette.types import ASGIApp

from .context import (
    TraceContext,
    generate_span_id,
    generate_trace_id,
    get_span_id,
    get_trace_id,
    set_span_id,
    set_trace_id,
)
from .models import Operations, ServiceName, TraceEvent, TraceStatus
from .publisher import TracePublisher


# Header names
TRACE_ID_HEADER = "X-Trace-ID"
SPAN_ID_HEADER = "X-Span-ID"


class TracingMiddleware(BaseHTTPMiddleware):
    """
    HTTP middleware for distributed tracing.

    Features:
    - Extracts trace_id from X-Trace-ID header or generates new one
    - Sets trace context for the request lifecycle
    - Publishes start/end events for API requests
    - Adds X-Trace-ID to response headers

    Example:
        from jarvis_tracing import TracingMiddleware, TracePublisher

        publisher = TracePublisher(redis_url="redis://localhost:6379")
        await publisher.connect()

        app.add_middleware(
            TracingMiddleware,
            publisher=publisher,
            service=ServiceName.BACKEND,
        )
    """

    def __init__(
        self,
        app: ASGIApp,
        publisher: TracePublisher,
        service: ServiceName = ServiceName.BACKEND,
        skip_paths: Optional[list[str]] = None,
    ):
        super().__init__(app)
        self.publisher = publisher
        self.service = service
        self.skip_paths = skip_paths or ["/health", "/metrics", "/favicon.ico"]

    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        # Skip tracing for certain paths
        if any(request.url.path.startswith(p) for p in self.skip_paths):
            return await call_next(request)

        # Extract or generate trace_id
        trace_id = request.headers.get(TRACE_ID_HEADER) or generate_trace_id()
        span_id = generate_span_id()

        # Set context for this request
        with TraceContext(trace_id) as ctx:
            start_time = time.perf_counter()

            # Build metadata
            metadata = {
                "method": request.method,
                "path": request.url.path,
                "query": str(request.query_params) if request.query_params else None,
                "client_host": request.client.host if request.client else None,
            }

            # Publish start event
            start_event = TraceEvent(
                trace_id=trace_id,
                span_id=ctx.span_id,
                service=self.service,
                operation=Operations.API_REQUEST,
                status=TraceStatus.STARTED,
                metadata=metadata,
                tags={
                    "endpoint": self._normalize_path(request.url.path),
                    "method": request.method,
                },
            )
            await self.publisher.publish(start_event)

            # Process request
            status_code = 500
            error_msg = None

            try:
                response = await call_next(request)
                status_code = response.status_code

                # Add trace headers to response
                response.headers[TRACE_ID_HEADER] = trace_id
                response.headers[SPAN_ID_HEADER] = ctx.span_id

                return response

            except Exception as e:
                error_msg = f"{type(e).__name__}: {str(e)}"
                raise

            finally:
                # Calculate duration
                duration_ms = int((time.perf_counter() - start_time) * 1000)

                # Update metadata
                metadata["status_code"] = status_code
                metadata["duration_ms"] = duration_ms

                # Determine status
                if error_msg or status_code >= 500:
                    status = TraceStatus.ERROR
                else:
                    status = TraceStatus.SUCCESS

                # Publish end event
                end_event = TraceEvent(
                    trace_id=trace_id,
                    span_id=ctx.span_id,
                    service=self.service,
                    operation=Operations.API_REQUEST,
                    status=status,
                    duration_ms=duration_ms,
                    metadata=metadata,
                    tags={
                        "endpoint": self._normalize_path(request.url.path),
                        "method": request.method,
                        "status_code": str(status_code),
                    },
                    error=error_msg,
                )
                await self.publisher.publish(end_event)

    def _normalize_path(self, path: str) -> str:
        """
        Normalize path for consistent tagging.

        /api/users/123 -> /api/users/{id}
        /api/tasks/abc-def -> /api/tasks/{id}
        """
        parts = path.split("/")
        normalized = []

        for part in parts:
            if not part:
                continue
            # UUID-like or numeric ID
            if len(part) > 8 and ("-" in part or part.isdigit()):
                normalized.append("{id}")
            else:
                normalized.append(part)

        return "/" + "/".join(normalized) if normalized else "/"


def extract_trace_headers(request: Request) -> tuple[Optional[str], Optional[str]]:
    """
    Extract trace headers from request.

    Returns (trace_id, span_id) tuple.
    """
    return (
        request.headers.get(TRACE_ID_HEADER),
        request.headers.get(SPAN_ID_HEADER),
    )


def inject_trace_headers(headers: dict) -> dict:
    """
    Inject current trace context into headers dict.

    Use when making outbound HTTP requests.
    """
    trace_id = get_trace_id()
    span_id = get_span_id()

    if trace_id:
        headers[TRACE_ID_HEADER] = trace_id
    if span_id:
        headers[SPAN_ID_HEADER] = span_id

    return headers
